File: int_matrix.cpp
//--------------------------------------------------------
//---------------Interaction matrix subroutine------------
//--------------------------------------------------------

#include <complex>
#include <iostream>
#include "nrutil_nr.h"
#include "nrtypes_nr.h"
#include <cmath>
#include <getopt.h>
#include <cstdio>
#include <cstdlib>
#include <fstream>

using namespace std;
using std::endl;
using std::ofstream;

void int_matrix(Vec_DP &k, double &mh, double &me, Vec_DP &nh, Vec_DP &ne, double &conc, double &Tempr, double &eps, int ind, string gway, Mat_DP &V){

//----------------Constantes------------------------------

//double h=1.05e-34;				// Planck constant
double e=1.602e-19;				// elementary charge
//double m0=9.1e-31;				// electron mass
double pi=3.14;				// pi	
double eps0=8.85e-12;				// dielectric constant
complex<double> ij(0,1);				// imaginary unit

//----------------Material parameters---------------------

double epss;				// dielectric constant

int j1;
int j2;
int j;

int l_k=k.size();		//length of k array
Vec_DP k1(k);		// k-space (2nd particle)
double phi[10002];
double q;

//----------------Definition of constants-----------------

if (ind==11){
	
	Vec_DP p(9);
	
	std::stringstream sstm;
	sstm << gway<<"f_factor.dat";	//valence band data reading
	string flnm=sstm.str();
	std::ifstream in(flnm.c_str(),std::ios::out);
	if(!in){  
		std::cout<<"Could not open file"<<sstm.str()<<std::endl;
		exit(0); 
	}
	j=0;
	while (!in.eof() && j<l_k) {
		in>>p[j];
		j++;
	}
	in.close();
}

for (j1=0;j1<3001;j1++){
	phi[j1]=j1*2*pi/3001;			// time array
	}
	
	ofstream outdata;
	outdata.open("/home/mk/my_projects/SBE_cpp/V.dat",ios::in|ios::trunc);
	

for (j1=0;j1<l_k;j1++){
    for (j2=0;j2<l_k;j2++){
        if ((j1==j2)||(j2==j1)){
            V[j1][j2]=0.0;
			for (j=1;j<3001;j++){
                q=sqrt(pow(k[j1],2)+pow(k1[j2],2)-2.0*k[j1]*k1[j2]*cos(phi[j]));	
				epss=1.0;
				V[j1][j2]=V[j1][j2]+(epss*e*e/(8.0*pow(pi,2)*eps*eps0*q))*fabs(phi[2]-phi[1]);
			}
		}
        else{
			V[j1][j2]=0.0;
			for (j=0;j<3001;j++){
                q=sqrt(pow(k[j1],2)+pow(k1[j2],2)-2.0*k[j1]*k1[j2]*cos(phi[j]));	
				epss=1.0;
				V[j1][j2]=V[j1][j2]+(epss*e*e/(8.0*pow(pi,2)*eps*eps0*q))*fabs(phi[2]-phi[1]);
			}
		}
		V[0][0]=0.0;
		outdata<<V[j1][j2]<<endl;
    }
}

outdata.close();

///* Open the file for writing. If it exists, append to it;
//otherwise, create a new file. */
//int fd = open ('/home/mk/science/Cpp_programms/SBE_project/src/V1.dat', O_WRONLY | O_CREAT | O_APPEND, 0666);
///* Write the timestamp to the file. */
//write(fd, V, l_k*l_k*sizeof(double));
///* All done. */
//close (fd);

//return 0;

}

File: int_matrix.f90
!--------------------------------------------------------
!---------------Interaction matrix subroutine------------
!--------------------------------------------------------

subroutine int_matrix(k,l_k,mh,me,nh,ne,conc,Tempr,eps,ind,gway,V)

implicit none

!----------------Constantes------------------------------

integer, parameter						:: dp = selected_real_kind(15, 307)
real :: h							! Planck constant
real :: e							! elementary charge
real :: m0							! electron mass
real :: pi=3.14						! pi	
real :: eps0						! dielectric constant
complex :: ij						! imaginary unit

!----------------Material parameters---------------------

real :: epss						! dielectric constant
real :: eps							! Permittivity

integer :: j1
integer :: j2
integer :: j

integer :: l_k						! length of k array
real, dimension(l_k) :: k			! k-space (1st particle)
real, dimension(l_k) :: k1			! k-space (2nd particle)
real, dimension(l_k,l_k) :: V		! Coulomb potential
real, dimension(10002) :: phi
double precision, dimension(9) :: p	! k-space (1st particle)
double precision :: q	
double precision :: f_f
real :: Tempr
real :: me
real :: mh
real :: ne
real :: nh
real :: conc

integer :: ind
character (len=100) :: gway
real(dp), dimension(9) :: ccc

!----------------Definition of constants-----------------

h=1.05e-34
e=1.602e-19
m0=9.1e-31
eps0=8.85e-12
ij=sqrt(CMPLX(-1.0))

if (ind==11) then
open(7,file=trim(gway)//'f_factor.dat')
read (7,'(e15.5)'), p
close(7)
endif

!call screening(l_k,k,conc,gway,ccc)

phi=(/(j1*2*pi/10001,j1=0,(10001))/)

k1=k

do j1=1,l_k
    do j2=1,l_k  
        if ((j1==j2).OR.(j2==j1)) then
            V(j1,j2)=0.0
	    do j=2,10001
	                    
!		call dielectrical_const(q,mh,me,nh,ne,conc,Tempr,epss)	
       		
                q=sqrt(k(j1)**2+k1(j2)**2-2.0*k(j1)*k1(j2)*cos(phi(j)))	
				f_f=p(1)*(q**8.0)+p(2)*(q**7.0)+p(3)*(q**6.0)+p(4)*(q**5.0)+p(5)*(q**4.0)+p(6)*(q**3.0)+p(7)*(q**2.0)+p(8)*q+p(9)
				!epss=ccc(1)+ccc(2)*q+ccc(3)*(q**2.0)+ccc(4)*(q**3.0)+ccc(5)*(q**4.0)+ccc(6)*(q**5.0)+ccc(7)*(q**6.0)+ccc(8)*(q**7.0)+ccc(9)*(q**8.0)
                !f_f=1.0
				call dielectrical_const(q,mh,me,nh,ne,conc,Tempr,epss)	
				V(j1,j2)=V(j1,j2)+f_f*(epss*(e**2.0)/(8.0*(pi**2.0)*eps*eps0*q))*abs(phi(2)-phi(1))
       		
	    enddo
        else
	    V(j1,j2)=0.0
	    do j=1,10001
                q=sqrt(k(j1)**2+k1(j2)**2-2.0*k(j1)*k1(j2)*cos(phi(j)))
				f_f=p(1)*(q**8.0)+p(2)*(q**7.0)+p(3)*(q**6.0)+p(4)*(q**5.0)+p(5)*(q**4.0)+p(6)*(q**3.0)+p(7)*(q**2.0)+p(8)*q+p(9)
				!epss=ccc(1)+ccc(2)*q+ccc(3)*(q**2.0)+ccc(4)*(q**3.0)+ccc(5)*(q**4.0)+ccc(6)*(q**5.0)+ccc(7)*(q**6.0)+ccc(8)*(q**7.0)+ccc(9)*(q**8.0)
                !f_f=1.0
				call dielectrical_const(q,mh,me,nh,ne,conc,Tempr,epss)	
				V(j1,j2)=V(j1,j2)+f_f*(epss*(e**2.0)/(8.0*(pi**2.0)*eps*eps0*q))*abs(phi(2)-phi(1))
	    enddo
        endif
    enddo
enddo

V(1,1)=0.0

open(7,file='v.dat')
write (7,'(400e15.7)'), V

return
end subroutine int_matrix
